tests/checkpointing/test_model_checkpoint.py

```diff
@@ -1248,3 +1248,42 @@ def test_trainer_checkpoint_callback_bool(tmpdir):
     mc = ModelCheckpoint(dirpath=tmpdir)
     with pytest.raises(MisconfigurationException, match="Invalid type provided for checkpoint_callback"):
         Trainer(checkpoint_callback=mc)
+
+
+def test_load_checkpoint_update_hparams(tmpdir):
+    class TestModel(BoringModel):
+        def __init__(self, x, y):
+            super().__init__()
+            self.x = x
+            self.save_hyperparameters()
+            self.y = y
+
+    model_checkpoint = ModelCheckpoint(dirpath=tmpdir, filename="{epoch:02d}")
+    trainer = Trainer(
+        max_epochs=1,
+        default_root_dir=tmpdir,
+        limit_train_batches=1,
+        limit_val_batches=1,
+        callbacks=[model_checkpoint],
+        logger=TensorBoardLogger(save_dir=tmpdir),
+        weights_summary=None,
+        progress_bar_refresh_rate=0,
+    )
+    model = TestModel(5, 6)
+    trainer.fit(model)
+
+    ckpt_path = os.path.join(tmpdir, "epoch=00.ckpt")
+    hparams_path = os.path.join(tmpdir, "default", "version_0", "hparams.yaml")
+    assert os.path.exists(ckpt_path)
+    assert os.path.exists(hparams_path)
+
+    # Make sure that even though y isn't saved in hparams_file,
+    # the model can still be loaded using the y value saved
+    # in the checkpoint.
+    model = TestModel.load_from_checkpoint(
+        checkpoint_path=ckpt_path,
+        hparams_file=hparams_path,
+    )
+
+    assert model.x == 5
+    assert model.y == 6
```

Comment:
```suggestion
def test_load_checkpoint_update_hparams(tmpdir):
    """
    Check that old hparams are updated when new hparams_file is provided
    """
    class TestModel(BoringModel):
        def __init__(self, x=None, y=None):
            super().__init__()
            self.save_hyperparameters()

    model_checkpoint = ModelCheckpoint(dirpath=tmpdir, filename="{epoch:02d}")
    trainer = Trainer(
        max_epochs=1,
        default_root_dir=tmpdir,
        limit_train_batches=1,
        limit_val_batches=1,
        callbacks=[model_checkpoint],
        logger=TensorBoardLogger(save_dir=tmpdir),
    )
    model = TestModel(5)
    trainer.fit(model)

    ckpt_path = os.path.join(tmpdir, "epoch=00.ckpt")
    hparams_path = os.path.join(tmpdir, "default", "version_0", "hparams.yaml")
    assert os.path.exists(hparams_path)
    assert model.hparams.x == 5
    assert model.hparams.y is None

    with open(hparams_path, 'w') as f:
        yaml.dump({'y': 10}, f)

    # Make sure that even though x isn't saved in new hparams_file,
    # the model can still be loaded using the x value saved
    # in the checkpoint.
    model = TestModel.load_from_checkpoint(
        checkpoint_path=ckpt_path,
        hparams_file=hparams_path,
    )

    assert model.hparams.x == 5
    assert model.hparams.y == 10
```
the test passes on master since save_hyperparameters saves x and y both in the checkpoint hparams. For this one, I think providing a new hparams file which only contains some missing hparams would be a better way to test this. You might need to import yaml up there. Also, I think this test is better suited for `tests/models/test_hparams`.

Reply:
Thanks for the feedback! However, I'm pretty sure `save_hyperparameters` would only save properties defined before it is called. I had verified locally (on PL `1.4.4`) and doing:

```
self.x = x
self.save_hyperparameters()
self.y = y
```

results in a `hparams.yaml` that only contains:

```
x: 5
```

Is this behavior changed in 1.5.0?

Regardless, I think your solution is better, just wanted to make sure my assumption about `save_hyperparameters` is correct?

Reply:
AFAIK `save_hyperparameters()` doesn't care what you save as a state variable and where you save it. It just captures all the args and kwargs and saves them. You can try reverting your changes in the codebase and run your current test, it will pass. I tried it on master and it's saving both x and y even when I don't attach them to `self`. You can check a few examples here: https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#save-hyperparameters. none of them assign it to the state but save_hyperparameters still saves them.